### PR TITLE
Create BatcherInterface

### PR DESCRIPTION
**Pull request by karupayun**
In this PR I'm creating BatcherInterface. Batcher will be divided between
this class and PrimitiveAssembler. BatcherInterface is the part that will
be overwritten by OpenGlBatcher to implement the functionality.

Discussed in go/stadia-orbit-testing-ui-elements, the interface contains:

- AddBox/Line/Triangle methods -> Inserting
- ResetElements() & GetNumElements()
- GetLayers() & DrawLayers() -> Drawing
- GetUserData() -> Mostly used for tooltips

And 2 trivial implementations:
- GetBatcherId()
- Push/PopTranslations() -> This is strange, I know.
a) we might be interested in using inheritance instead of composition here?
b) Should we instead just rename BatcherInterface?
c) Should we moved this to every implementation?
d) Should we leave this as it is? I'm leaning toward this one but I'm open
to opinions.

Next step is linking OpenGlBatcher to BatcherInterface and renaming Batcher
to PrimitiveAssembler (Merged it in my local branch: https://github.com/karupayun/orbit/commit/1d4d264fc5f1f90c033b62bc2d0b22f45750401f)

http://b/225173189

### CODE REVIEW

File: src/OrbitGl/BatcherInterface.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+#ifndef ORBIT_GL_BATCHER_INTERFACE_H_
+#define ORBIT_GL_BATCHER_INTERFACE_H_
+
+#include "ClientProtos/capture_data.pb.h"
+#include "Geometry.h"
+#include "PickingManager.h"
+#include "TranslationStack.h"
+
+using TooltipCallback = std::function<std::string(PickingId)>;
```
[the remark this answers is not in the document]
Comment (antonrohr) [reply src/OrbitGl/BatcherInterface.h:13]: Maybe even in `PickingUserData` ?

**Review by antonrohr — COMMENTED** [review]
In general, LGTM, but I would like someone else to also take a look since I am not super familiar. 

Regarding it being BatcherInterface and the batcher_id and translations: I think the name `Batcher` is free after the rename to OpenGlBatcher and PrimitiveAssembler, so one option would be to just add another inheritance layer. 
```
class BatcherInterface {};
class Batcher : public BatcherInterface {}; // Only adds the batcher_id and translations
class OpenGlBatcher: public Batcher {};
```
This is just an idea, and I am not sure if this actually works in regard to mocking etc. I am also fine with all other solutions

**Comment by karupayun** [comment]
> In general, LGTM, but I would like someone else to also take a look since I am not super familiar.
> 
> Regarding it being BatcherInterface and the batcher_id and translations: I think the name `Batcher` is free after the rename to OpenGlBatcher and PrimitiveAssembler, so one option would be to just add another inheritance layer.
> 
> ```
> class BatcherInterface {};
> class Batcher : public BatcherInterface {}; // Only adds the batcher_id and translations
> class OpenGlBatcher: public Batcher {};
> ```
> 
> This is just an idea, and I am not sure if this actually works in regard to mocking etc. I am also fine with all other solutions

I like the idea so the Interface is a real interface. I added a TODO (https://b/227302285), since I planned to do this change first, after renaming Batcher and therefore this fixup will come later.

**Review by antonrohr — APPROVED** [review]
LGTM, 

maybe it would be better if someone else also looks at this, but I am also fine with it being merged now. If something needs to be changed, we can always change it

File: src/OrbitGl/BatcherInterface.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+//
+// By calling BatcherInterface::AddXXX, primitives are added to internal CPU buffers, and sorted
+// into layers formed by equal z-coordinates. Each layer should then be drawn seperately with
+// BatcherInterface::DrawLayer(). BatcherInterface also provides a method to get the user data
+// given a picking_id (in general used for tooltips).
+//
+// BatcherInterface is a pure interface class except for Translations and BatcherId.
+class BatcherInterface {
+ public:
+  BatcherInterface(BatcherId batcher_id) : batcher_id_(batcher_id) {}
```
Comment (beckerhe) [inline src/OrbitGl/BatcherInterface.h:36]: `explicit`

File: src/OrbitGl/BatcherInterface.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+// Collects primitives to be rendered at a later point in time.
+//
+// By calling BatcherInterface::AddXXX, primitives are added to internal CPU buffers, and sorted
+// into layers formed by equal z-coordinates. Each layer should then be drawn seperately with
+// BatcherInterface::DrawLayer(). BatcherInterface also provides a method to get the user data
+// given a picking_id (in general used for tooltips).
+//
+// BatcherInterface is a pure interface class except for Translations and BatcherId.
+class BatcherInterface {
```
Comment (beckerhe) [inline src/OrbitGl/BatcherInterface.h:34]: The class is missing a defaulted virtual destructor.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

**Pull request merged**